core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -366,14 +368,18 @@ public JoinReduceExpressionsRule(Class<? extends Join> joinClass,
 
     @Override public void onMatch(RelOptRuleCall call) {
       final Join join = call.rel(0);
-      final List<RexNode> expList = Lists.newArrayList(join.getCondition());
+      final RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+      final RexSimplify rexSimplify =
+          new RexSimplify(rexBuilder, RelOptPredicateList.EMPTY, EXECUTOR);
+      final RexNode condition = rexSimplify.eliminateCommonExprInCondition(join.getCondition());
+
+      final List<RexNode> expList = Lists.newArrayList(condition);
```

Comment:
Is there a reason `eliminateCommonExprInCondition` is performed before regular simplification?
Should `RelOptPredicateList` be taken into account as well?

Reply:
Cuz this opt can simplify condition, so put it before regular simplification seems intuitive.

Reply:
Regular simplification might simplify AND operands and it might enable common expression factoring.

Have you considered the possibility to integrate `factor common subexpression` into a regular `simplify(..)` call?

Reply:
Do you mean to add this optimization to RexSimplify's main process? Yes, we can do that, but currently, use this to simplify "join cond" has the most benefit. Maybe we can add this to RexSimplify's main process next time, I'll find more scenario about this optimization.

Reply:
@vlsi there was a discussion about where to put this new optimization. @Aaaaaaron is following @julianhyde 's proposal to take a conservative approach and include this new simplification, for the moment, only in the join push down condition rule (see https://issues.apache.org/jira/browse/CALCITE-4375?focusedCommentId=17226397&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17226397)

Reply:
I agree with @julianhyde that "CNF rewrite only in the particular rule".

However, I believe it should be implemented as a flag in RexSimplify that would optionally enable "CNF rewrite".
The rules that need CNF rewrite would set the flag.

The current implementation leaves the door open for the fight between different simplifications: the regular simplifier might want to simplify in its own way (e.g. ReduceExpressionsRule), and JoinConditionSimplify rule might want to rewrite into CNF.

Reply:
@Aaaaaaron I think this idea mentioned by @vlsi makes sense: integrating this new simplification into RexSimplify via a new optional flag (disabled by default, enable in cases like this one). WDYT?

Reply:
I don’t like the idea of RexSimplify having “modes”. Something is either simple or it isn’t. If we can’t agree on what is simple, then it isn’t simple.

I think the CNF/DNF rewrites are a better pattern to follow.

Reply:
So @julianhyde you agree with the current status of the PR? At the current state, this extra simplification is performed apart from RexSimplify, in the relevant rule.

Reply:
The current PR is very unclear.

It doesn't state which component is being improved. I guess it's JoinReduceExpressionsRule?

It adds a method to RexSimplify which is orthogonal to the other purposes of RexSimplify. That's like adding a method to StringBuilder to also tell you whether it's going to rain today.

There are changes to parameter types of public methods. Let's make sure that is not breaking.

Changing to use `HashSet` instead of `List` will cause optimization to become nondeterministic. One option is to switch to `LinkedHashSet`. But it is rather expensive. You should seriously consider continuing to use lists but just ensure that they are unique when you build them. This pattern works well:

```
final Set<String> set = new HashSet<>();
final List<String> list = new ArrayList<>();
for (s in source()) {
  if (set.add(s)) {
    list.add(s);
  }
}
return list;
```

Reply:
Hi @julianhyde 
The improvement as I comment in JIRA, both have ut in this PR:
SQL:
```
SELECT * FROM emps,depts
WHERE
(emps.name = depts.name AND empno=1)
OR
(emps.name = depts.name AND empno=2)
 ```

And the join after optimizer is:

`EnumerableNestedLoopJoin(condition=[OR(AND(=($1, $11), =($0, 1)), AND(=($1, $11), =($0, 2)))], joinType=[inner])`

 

In fact ($1, $11) can be extracted, and the join can be:

`HashJoin(condition=[AND(=($1, $11), OR(=($0, 1), =($0, 2)))], joinType=[inner]) `

 

We found 8 queries like this patten in TPC-DS, he benefits of optimization are great:

1. EnumerableNestedLoopJoin -> HashJoin
2. filter can be push down

Reply:
To use "set" here is a performance improvement when you want to find the common part of the expressions(`retainAll `),  I think for the deterministic optimization result, we can use `LinkedHashSet `, it will not be so big that has thousands elements.